poster/src/poster.ts

```diff
@@ -2,224 +2,173 @@ import { postWithRetries } from './post_with_retries';
-      if (!supportedAssets.includes(asset.toUpperCase())) {
-        index++;
-        continue;
-      }
-
-      const prev_price = await helpers.getPreviousPrice(sourceAddress, asset, dataAddress, web3);
-      console.log(`For asset ${asset}: prev price = ${prev_price}, new price = ${price}`);
-
-      // Update price only if new price is different by more than delta % from the previous price
-      if (!inDeltaRange(delta, Number(price), prev_price)) {
-        console.log(`Not in delta range for asset ${asset}, price = ${Number(price)}, prev price = ${prev_price}`);
-        filteredPrices[asset] = price;
-        filteredMessages.push(payload.messages[index]);
-        filteredSignatures.push(payload.signatures[index]);
-      }
-      index++;
-    }
-    payload.prices = filteredPrices;
-    payload.messages = filteredMessages;
-    payload.signatures = filteredSignatures;
-  }))
-
-  // Filter payloads with no updated prices
-  const filteredPayloads = payloads.filter(payload => payload.messages.length > 0);
-  return filteredPayloads;
+export async function filterPayloads(
+    payloads: OpenPriceFeedPayload[],
+    viewAddress: string,
+    supportedAssets: string[],
+    delta: number,
+    web3: Web3): Promise<OpenPriceFeedItem[]> {
+
+  const dataAddress = await getDataAddress(viewAddress, web3);
+
+  let filteredFeedItems = await Promise.all(payloads.map(async payload => {
+    return await Promise.all(zip(payload.messages, payload.signatures).map(([message, signature]) => {
+        const {
+          dataType,
+          timestamp,
+          symbol,
+          price
+        } = decodeMessage(message, web3)
+
+        return {
+          message,
+          signature,
+          dataType,
+          timestamp,
+          symbol: symbol.toUpperCase(),
+          price: Number(price)
+        };
+      }).filter(({message, signature, symbol}) => {
+        return supportedAssets.includes(symbol.toUpperCase());
+      }).map(async (feedItem) => {
+        const source = await getSourceAddress(dataAddress, feedItem.message, feedItem.signature, web3);
+        const prev = await getPreviousPrice(source, feedItem.symbol, dataAddress, web3);
+
+        return {
+          ...feedItem,
+          source,
+          prev: Number(prev) / 1e6
+        };
+      })).then((feedItems) => {
+        return feedItems.filter(({message, signature, symbol, price, prev}) => {
+          return !inDeltaRange(delta, price, prev);
+        });
+      });
+  }));
+
+  let feedItems = filteredFeedItems.flat();
+
+  feedItems
+    .forEach(({source, symbol, price, prev}) => {
+      console.log(`Setting Price: source=${source}, symbol=${symbol}, price=${price}, prev_price=${prev}`);
+    });
+
+  return feedItems;
 }
 
 // Checks if new price is less than delta percent different form the old price
-function inDeltaRange(delta: number, price: number, prev_price: number) {
+// Note TODO: price here is uh... a number that needs to be scaled by 1e6?
```

Comment:
Yes, the price from coinbase api needs to be scaled to 1e6, it's coming in this format:
"BTC": "9212.115", "ETH": "238.03", "XTZ": "3.04975" ...
and prevPrice is scaled to1e6 already.

Reply:
Yup, but I think the answer is: why is this code written that it accepts two numbers but assumes they are of different scales? I fixed this by descaling the price before this function is called.